src/goto-instrument/contracts/dynamic-frames/dfcc_instrument.h

```diff
@@ -34,6 +41,48 @@ class cfg_infot;
 class dfcc_libraryt;
 class dfcc_utilst;
 
+/// A local pragma used to keep track instructions in loops.
+const char CONTRACT_PRAGMA_loop_id[] = "contracts:loop-id";
+const char CONTRACT_PRAGMA_loop_body[] = "contracts:loop-id-body";
+const char CONTRACT_PRAGMA_loop_head[] = "contracts:loop-id-head";
+const char CONTRACT_PRAGMA_loop_latch[] = "contracts:loop-id-latch";
+const char CONTRACT_PRAGMA_loop_exiting[] = "contracts:loop-id-exiting";
+
+const char empty_pragma[] = "";
+
+// A graph node type that stores information about a loop.
+// We create a DAG representing nesting of various loops in goto_function,
+// sort them topologically, and instrument them in the top-sorted order.
+//
+// The goal here is not avoid explicit "subset checks" on nested write sets.
+// When instrumenting in a top-sorted order,
+// the outer loop would no longer observe the inner loop's write set,
+// but only corresponding `havoc` statements,
+// which can be instrumented in the usual way to achieve a "subset check".
```

Comment:
this was how it used to work, but is not true in this implementation. Here we have explicit inner-outer loop and function-top-level-loop inclusion checks, and we tag all instructions added to instrument a given loop so they don't get instrumented again for the outer loop;